source/common/http/http1/header_formatter.cc

```diff
@@ -0,0 +1,22 @@
+#include "common/http/http1/header_formatter.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http1 {
+std::string ProperCaseHeaderKeyFormatter::format(absl::string_view key) const {
```

Comment:
Looking at https://en.wikipedia.org/wiki/List_of_HTTP_header_fields

This won't do A-IM, TE Content-MD5, HTTP2-Setting, and about 10 others the way an HTTP/1 user might expect.

I'd lean towards saying we do standard capitalization of standard HTTP 1.1 headers that Envoy knows about (e.g. TE) with a link to headers Envoy knows about.  ProperCaseExceptForRegisteredStandardizedHeaderExceptions is too wordy, but I think that's what we want?

This way if someone later on desperately needs P3P cased they awy they want, they can just add it to the list of Envoy known headers.  And when we eventually add the Envoy extension mechanism for arbitrary headers we keep talking about, folks can do their own custom builds with TheirWhackyCAPs if they want.

Alternately if we want to go for the simple case, I think we should document there are known standard headers which will still look unusual to HTTP/1 consumers, and someone can implement ProperCaseByDefaultButWithOverrides later now we have an API for things.

@mattklein123 WDYT?

Reply:
Maybe `StandardOrProperCase`? I can go either way on whether we want to handle well known headers, it sounds a tad bit more in line with what someone might expect but it might be overkill.

Looking at Go's net/http (https://golang.org/src/net/http/header.go#L214) they do the same kind of formatting this PR does when canonicalizing headers except that they only uppercase letters after a hyphen.

Reply:
I don't have a strong opinion here as long as it is well documented. The way that we have things set up at the API layer we can always make things configurable later if needed.

Reply:
A soft suggestion of having `StandardOrProperCase` behavior; and perhaps condensing the config name down to `StandardCase` (which would follow well-known casings for well-known headers, and standard casing patterns for H/1 otherwise).